docs/kubernetes/mailu/front.yaml

```diff
@@ -1,23 +1,41 @@
-
-apiVersion: extensions/v1beta1
-kind: Deployment
+apiVersion: apps/v1beta2
+kind: DaemonSet
```

Comment:
Why a `DaemonSet`? I believed the use of `Deployment` before was good. Also see my notes below WRT affinity.

Reply:
The `Deployment` usage has it limitations, when the container starts it opens up node ports (all mail ports) and whenever a node fails or the pod fails it can be rescheduled to another node which means you need to adapt DNS settings every time your pod reschedules. The only solution to this is to use a cloud based LoadBalancer which connects to a kubernetes service (which limits usage to specific cloud providers only)
With the `DaemonSet` configuration, the front container is placed on the same level as the Nginx Ingress Controller, meaning every node is able to accept mail connections. This keeps the overview of the complete system transparent. You have the Ingress Controller DaemonSet opening up ports 80 and 443 (accepting all kinds of HTTP(S) services you have on your cluster) and the Mailu Front DaemonSet accepting all mail ports.
That's why I still go for the `DaemonSet` configuration.
Hopefully this clears things out

Reply:
I see, that's an innovative solution to the problem. Most of the solutions I have seen consider an internal load balancer (HAProxy-based solutions are good for this) such that there is a singular endpoint for the services as they are distributed within the cluster. This solves the ephemeral (portable) pod problem. The internal load balancer is then exposed externally using whatever facility is available, including MetalLB for bare metal (non-SaaS) clusters.

That said, there are a lot of moving parts with this common pattern when the choice of CNI is included and in the real world and the lack of a comprehensive TCK for kubernetes means that these moving parts often don't work together they way they should, which is very frustrating for users.

In my deployment, I have many addresses that are not associated directly with interfaces. This is facilitated by MetalLB announcing these addresses to upstream BGP and service routed through Cilium CNI. So `NodePort` and `hostPort` are not viable solutions.

Reply:
@hacor I also see that the conversation does not leave it's state of "resolved" after leaving another comment. Hopefully it is not hidden on your end.